### PR TITLE
Fix Mini App exercise flow: empty message text and sendData keyboard type

### PR DESCRIPTION
**Pull request by yogson**
## Summary

Two bugs prevented the vocabulary flashcard Mini App from working end-to-end:

- **Empty message text** — `build_webapp_button_channel_response()` produced a `ChannelResponse` with blank `text`, causing Telegram to reject it with _"message text is empty"_. The orchestrator handler now uses the tool result's `message` field as fallback text via a new `pending_webapp_message` field on `OrchestratorResult`.
- **`sendData()` silently failing** — `Telegram.WebApp.sendData()` only works when the Mini App is opened from a **reply-keyboard** button (`KeyboardButton` + `WebAppInfo`). Opening via an inline-keyboard button gives the Mini App no `sendData` capability. Changed `ui_kind` to `reply_keyboard` and extended egress rendering to attach `WebAppInfo` when `action.web_app_url` is set.

Additional changes included in this PR:
- New `process_exercise_results` tool registered in config and implemented
- `web_app_data` Telegram messages now handled in ingress (builder + service routing)
- `VocabularyStore` wired into orchestrator lifespan
- `start_exercise` tool result now includes an `actions` list with the WebApp button
- Language-learning concept doc added under `docs/`

## Changes by area

| Area | Files |
|---|---|
| WebApp button extraction | `src/assistant/agent/webapp_button_extractor.py` (new) |
| Orchestrator result model | `src/assistant/core/orchestrator/models.py`, `service.py` |
| API handler routing | `src/assistant/api/orchestrator_handler.py`, `utils.py` |
| Telegram egress (keyboard rendering) | `src/assistant/channels/telegram/egress.py`, `models.py` |
| Telegram ingress (web_app_data) | `src/assistant/channels/telegram/ingress_builders.py`, `ingress_service.py` |
| Exercise tools | `start_exercise.py`, `process_exercise_results.py` (new) |
| Config | `config/capabilities/language_learning.yaml`, `config/tools.yaml` |
| Tests | `tests/assistant/channels/telegram/test_egress.py` |

## Test plan

- [x] All 1 055 existing tests pass (`make test`)
- [x] New test `test_reply_keyboard_with_webapp_url_uses_webapp_info` verifies `WebAppInfo` is attached to the reply-keyboard button
- [ ] Manual smoke test: send `/exercise` → bot responds with reply-keyboard button → open Mini App → complete exercise → `sendData()` delivers results back to bot

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/assistant/agent/webapp_button_extractor.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+"""
+
+from typing import Any
+
+from pydantic_ai.messages import ModelMessage, ModelRequest, ToolReturnPart
+
+from assistant.agent.message_converters import _parse_tool_result_content
+
+_START_EXERCISE_TOOL_NAME = "start_exercise"
+
```
Comment (yogson) [inline src/assistant/agent/webapp_button_extractor.py:15]: **Dead code — `_START_EXERCISE_TOOL_NAME` is never used.**

This constant is defined but never referenced anywhere in the file. The extractor currently accepts webapp buttons from *any* tool result that has `actions[].web_app_url`, which is generic and forward-compatible, but the name suggests it was meant to narrow the scope to `start_exercise` results only.

Two options:
1. **Use it as a filter** — add a guard `if req_part.tool_name != _START_EXERCISE_TOOL_NAME: continue` before parsing, preventing accidental pickup of webapp buttons from future tools that may also return `actions`.
2. **Remove it** — if the generic behaviour is intentional, delete the constant to avoid confusing future readers.

Either way the dead constant should not remain as-is.

File: src/assistant/api/orchestrator_handler.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import structlog
 from pydantic_ai import ModelHTTPError, UnexpectedModelBehavior
 
-from assistant.api.utils import build_text_channel_response
+from assistant.api.utils import build_text_channel_response, build_webapp_button_channel_response
 from assistant.channels.telegram import ChannelResponse, NormalizedEvent, TelegramAdapter
```
Comment (yogson) [inline src/assistant/api/orchestrator_handler.py:9]: **Latent edge case: empty `text` can still reach Telegram.**

```python
webapp_text = response_text.strip() or orch_result.pending_webapp_message or ""
```

The final `or ""` fallback means an empty string can still be passed to `build_webapp_button_channel_response`. The original bug was precisely an empty `text` causing Telegram API rejection (`Bad Request: message text is empty`). Today `start_exercise` always returns a non-empty `message`, so this is safe in practice — but it is a latent trap for any future tool that returns webapp buttons without a message field.

Consider replacing the silent fallback with a runtime guard:

```python
webapp_text = response_text.strip() or orch_result.pending_webapp_message or ""
if not webapp_text:
    logger.error("webapp_button_response: empty text, cannot send to Telegram")
    # fall through to normal text response or raise a descriptive error
```

A descriptive `ValueError` would make the root cause immediately visible rather than surfacing as an opaque Telegram API error downstream.

File: src/assistant/agent/webapp_button_extractor.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+def _extract_pending_webapp_buttons(
+    new_messages: list[ModelMessage],
```
Comment (yogson) [inline src/assistant/agent/webapp_button_extractor.py:18]: **Missing unit tests for `_extract_pending_webapp_buttons`.**

This function is the glue between raw pydantic-ai tool results and the `OrchestratorResult` that drives the channel response. There are no tests for it. Suggested cases (following the existing pattern in `tests/assistant/agent/`):

- **(a) Happy path** — a `ModelRequest` with a `ToolReturnPart` whose content includes `actions[{"web_app_url": "https://..."}]` and a `message` field; assert the returned tuple matches.
- **(b) Actions present, no `web_app_url`** — `actions` list exists but no entry has `web_app_url`; assert `None` is returned.
- **(c) No `actions` key** — `ToolReturnPart` content has no `actions` field; assert `None` is returned.
- **(d) No `ToolReturnPart` at all** — `new_messages` contains only `ModelResponse` parts; assert `None` is returned.

Without these tests, a refactor of `_parse_tool_result_content` or a schema change in the tool result could silently break the exercise flow.

File: src/assistant/extensions/language_learning/tools/process_exercise_results.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    CardDirection,
+    ExerciseResultPayload,
+    LearningStatus,
+)
+
+logger = structlog.get_logger(__name__)
+
+
+async def process_exercise_results(
+    ctx: RunContext[TurnDeps],
```
Comment (yogson) [inline src/assistant/extensions/language_learning/tools/process_exercise_results.py:27]: **Missing unit tests for `process_exercise_results`.**

The existing `test_start_exercise.py` sets a good precedent; this new tool needs the same treatment. Suggested cases:

- **(a) Valid payload, happy path** — valid JSON with 3 results; assert `status == "ok"`, `updated_count`, and `status_breakdown` are correct.
- **(b) Invalid JSON** — pass a non-JSON string; assert `status == "rejected_invalid"` and `reason` mentions the parse error.
- **(c) Schema mismatch** — valid JSON but missing required fields (e.g., no `results` key); assert `status == "rejected_invalid"`.
- **(d) Empty results list** — `{"type": "exercise_results", "results": []}` — assert `status == "rejected_invalid"`.
- **(e) Store unavailable** — `ctx.deps.vocabulary_store is None`; assert `status == "unavailable"`.
- **(f) Store raises exception** — mock `store.process_exercise_results` to raise; assert `status == "error"`.

Without these, regressions in the SM-2 update path or JSON parsing are not caught by CI.

File: src/assistant/channels/telegram/ingress_builders.py [hunk 5]
```diff
@@ -150,6 +150,32 @@ def build_media_group_event(
     )
 
 
+def build_web_app_data_event(
+    message: dict[str, Any],
```
Comment (yogson) [inline src/assistant/channels/telegram/ingress_builders.py:154]: **Missing ingress tests for `build_web_app_data_event` and `web_app_data` routing.**

`build_web_app_data_event` has no test. This function (and the routing in `ingress_service.py` that calls it) forms the entire return path from the Mini App back to the agent — if it silently misroutes or drops the payload the user gets no feedback after completing a flashcard session.

Suggested tests (can live alongside the existing ingress builder tests):

- **`build_web_app_data_event` basic** — assert that `web_app_data.data` maps to `event.text`, `event_type` is `USER_TEXT_MESSAGE`, and `idempotency_key` is formatted correctly.
- **`build_web_app_data_event` with empty `data`** — assert `event.text is None` (the `data_text or None` guard).
- **`_normalize_message` routing** — given a message dict with a `web_app_data` key, assert that the `web_app_data` branch fires (not the plain-text fallback). This is important because the `elif "web_app_data" in message` branch must take priority over the default `build_text_event` path.

**Review by yogson — COMMENTED** [review]
## Review Summary

The two core bugs are correctly identified and fixed. The `ui_kind="reply_keyboard"` + `WebAppInfo` approach is exactly right per the Telegram Bot API docs — `sendData()` is only available when a Mini App is launched from a reply-keyboard button, and the previous inline-keyboard wiring silently broke it. The `pending_webapp_message` fallback in the orchestrator cleanly addresses the empty `text` rejection from Telegram. The new egress test for `WebAppInfo` on reply-keyboard buttons is solid.

Before merging, there are four issues that need attention: one dead constant, one latent edge case in the empty-text guard, and two missing test suites.
